File: 1286/Motoboy.cpp
using namespace std;
#include <iostream>
#include <vector>
#include<bits/stdc++.h>
#include <algorithm>

int knapsack(int W, int b[], int wt[], int n){
    
    
    int V[n+1][W+1];
    
    for (int i = 0; i<=W; i++){
        V[0][i] = 0;
    }
    for (int i = 1; i<=n; i++){
        V[i][0] = 0;
    }
    
    for (int i = 1; i <= n; i++){
        for (int j = 1; j<=W; j++){
            if (wt[i-1] <= j){
                if ((b[i-1] + V[i-1][j-wt[i-1]])>V[i-1][j]){
                    V[i][j] =b[i-1] + V[i-1][j-wt[i-1]];
                }else{
                    V[i][j] = V[i-1][j];
                }
            }else{
                V[i][j] = V[i-1][j];
            }
        }
    }
    
    return V[n][W];
    
}
int main(){
    int n, W, t, q;
    
    cin >> n; cin.ignore();
    int b[n];
    int w[n];
    while (n!=0){
        
        cin >> W; cin.ignore();
        
        for (int i = 0; i < n; i++){
            cin >> t >> q;
            b[i] = t;
            w[i] = q;
        }
        
        int value = knapsack(W,b,w,n);
    
        cout << value << " min." << endl;
        cin >> n;
        
    }
    
    
}

File: 1286/Motoboy.py
def knapsack(W, b, w):
    V = []
    
    for i in range(0,len(b)+1):
        V.append([-1]*(W+1))
    for i in range(0,W+1):
        V[0][i] = 0
    for i in range(1,len(b)+1):
        V[i][0] = 0
    
    for i in range(1,len(b)+1):
        for j in range(1, W+1):
            if w[i-1] <= j:
                if b[i-1] + V[i-1][j-w[i-1]] > V[i-1][j]:
                    V[i][j] = b[i-1] + V[i-1][j-w[i-1]]

                else:
                    V[i][j] = V[i-1][j]
            else:
                V[i][j] = V[i-1][j]
   
    return V[len(b)][W]

n = int(input())


while n!=0:
    

    W = int(input())
    b = []
    w = []

    for i in range(n):
        t, q = map(int, input().split())

        b.append(t)
        w.append(q)


    value = knapsack(W, b, w)
    print("%s min."%(value))
    n = int(input())
